pkg/apis/messaging/v1alpha1/pipeline_types.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/pkg/apis"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"github.com/knative/pkg/webhook"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type Pipeline struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Pipeline.
+	Spec PipelineSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Pipeline. This data may be out of
+	// date.
+	// +optional
+	Status PipelineStatus `json:"status,omitempty"`
+}
+
+// Check that Pipeline can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Pipeline)(nil)
+var _ apis.Defaultable = (*Pipeline)(nil)
+var _ apis.Immutable = (*Pipeline)(nil)
+var _ runtime.Object = (*Pipeline)(nil)
+var _ webhook.GenericCRD = (*Pipeline)(nil)
+
+type PipelineSpec struct {
+	// Steps is the list of Subscribers (processors / functions) that will be called in the order
+	// provided.
+	Steps []eventingv1alpha1.SubscriberSpec
+
+	// TODO: Specify the CRD Channel which should be used to create the underlying Channels.
+
+	// Subscriber is the addressable that optionally receives replies from the last step in the pipeline.
+	// +optional
+	Subscriber *eventingv1alpha1.SubscriberSpec `json:"subscriber,omitempty"`
+}
+
+// PipelineStatus represents the current state of a Pipeline.
+type PipelineStatus struct {
+	// inherits duck/v1alpha1 Status, which currently provides:
+	// * ObservedGeneration - the 'Generation' of the Service that was last processed by the controller.
+	// * Conditions - the latest available observations of a resource's current state.
+	duckv1alpha1.Status `json:",inline"`
+
+	// SubscriptionStatuses is an array of corresponding Subscription statuses.
+	// Matches the Spec.Steps array in the order.
+	SubscriptionStatuses []eventingv1alpha1.SubscriptionStatus
+
+	// ChannelStatuses is an array of corresponding Channel statuses.
+	// Matches the Spec.Steps array in the order.
+	ChannelStatuses []eventingv1alpha1.ChannelStatus
```

Comment:
This will include stuff like the address of each intermediate Channel. I'm guessing we don't want that. How about

```golang
type PipelineChannelStatus struct {
  Channel v1.ObjectReference
  ReadyCondition duckv1alpha1.Condition
}
```

Reply:
Great idea!